### PR TITLE
sql: add option to use lookup join for db-scope virtual tables

### LINKED ISSUE #90115: sql: DB-scope virtual tables are slow despite a small DB
**Is your feature request related to a problem? Please describe.**

CockroachDB offers per-database metadata schemas `information_schema` and `pg_catalog`. The tables in these schemas are implemented "virtually"; instead of being real tables, we run some code over the internal metadata structures to generate the output. Today, in order to populate the correct set of descriptors, we unconditionally fetch all descriptors. In cases where individual databases are not large, but the total number of descriptors is large, virtual tables are shockingly slow. Importantly, these virtual tables don't need to look at most of these descriptors.

**Describe the solution you'd like**
We can avoid looking at the unwanted descriptors by performing a lookup first against the namespace table and then fetching the descriptors in the database.

**Additional context**
This approach won't help with cases like https://github.com/cockroachdb/cockroach/issues/83626 where individual schemas or databases have large numbers of objects. 

Epic: CRDB-20865

Jira issue: CRDB-20592

### PR DESCRIPTION
**Pull request by ajwerner**
information_schema and pg_catalog are always scoped to a database. Before this change, we would always look up all descriptors to materialize virtual tables, even if we only care about descriptors in the current database. The new approach: scan the namespace table to find the relevant descriptors and then only resolve those.

On my local machine where I loaded up a few megabytes of descriptors into 6k databases, a simple query like `SELECT * FROM information_schema.sequences` takes 1.2s without the setting enabled and 1ms with it.

Fixes #90115

Epic: CRDB-20865
    
Release note (performance improvement): Tables in `pg_catalog` and
`information_schema` (when not explicitly referenced as
`"".information_schema`) may now be much faster if the current database has
a small number of relations relative to the total number in the cluster.

### CODE REVIEW

File: pkg/sql/catalog/descs/collection.go [hunk 1]
```diff
@@ -364,6 +366,81 @@ func newMutableSyntheticDescriptorAssertionError(id descpb.ID) error {
...
+) (nstree.Catalog, error) {
+	ns, err := tc.stored.GetAllDescriptorNamesForDatabase(ctx, txn, db)
+	if err != nil {
+		return ns, err
+	}
+	var ids catalog.DescriptorIDSet
+	ids.Add(db.GetID())
+	_ = ns.ForEachNamespaceEntry(func(e nstree.NamespaceEntry) error {
+		if !strings.HasPrefix(e.GetName(), catconstants.PgTempSchemaName) &&
+			e.GetID() != 29 {
```
Comment (chengxiong-ruan) [inline pkg/sql/catalog/descs/collection.go:382]: 29 -> `catconstants.PublicSchemaID`

File: pkg/sql/schema_resolver.go [hunk 2]
```diff
@@ -300,12 +300,15 @@ func (sr *schemaResolver) GetTypeDescriptor(
 	ctx context.Context, id descpb.ID,
 ) (tree.TypeName, catalog.TypeDescriptor, error) {
 	tc := sr.descCollection
-	desc, err := tc.GetImmutableTypeByID(ctx, sr.txn, id, tree.ObjectLookupFlags{})
+	flags := sr.CommonLookupFlagsRequired()
+	flags.ParentID = sr.typeResolutionDbID
```
Comment (chengxiong-ruan) [inline pkg/sql/schema_resolver.go:304]: Thanks for doing this, I'd have forgotten about doing this for types if it's me.

**Review by chengxiong-ruan — COMMENTED** [review]
Looking good! Do we want to backport this?

**Review by ajwerner — COMMENTED** [review]
> Looking good! Do we want to backport this?

I'd like to try. There's a customer who would like it. I don't think 22.1 will be clean. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-:-NFAqz0Q9kVD9n4GM2pi:b-cuhrru)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan)

___
*[`pkg/sql/catalog/descs/collection.go` line 382 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NF5Sc-QiDGLShcvgkHY-r3-382:-NFAqtCNBG3sgoLlc64g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dd2396e771a4488095f16e38ea3597fed2653ff/pkg/sql/catalog/descs/collection.go#L382)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

29 -> `catconstants.PublicSchemaID`
</blockquote></details>

Done.

**Review by chengxiong-ruan — APPROVED** [review]
LGTM

**Comment by chengxiong-ruan** [comment]
> > Looking good! Do we want to backport this?
> 
> I'd like to try. There's a customer who would like it. I don't think 22.1 will be clean.

Yeah...I bet the backport would be annoying...Thanks again for working on this, huge.

**Comment by ajwerner** [comment]
Yeah, I'm giving up some hope on 22.1. 22.2 should be fine.

**Comment by ajwerner** [comment]
@rafiss do you want to give this a look?

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90116)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @chengxiong-ruan)

___
*[`-- commits` line 4 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFJREG-rwTRfqrYmsS):*
not entirely accurate: we support queries like `select * from "".information_schema.tables`
___
*[`.ijwb/.idea/modules.xml` line 1 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFF9lqAExNcG8_puDkP:-NFF9lqAExNcG8_puDkQ:baw2zaw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ba39c4a53c20ba112715166de8f01cf0b98244/.ijwb/.idea/modules.xml#L1)):*
> ```XML
> <?xml version="1.0" encoding="UTF-8"?>
> ```

was this committed on purpose?
___
*[`pkg/sql/crdb_internal.go` line 517 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFJBaB10UhzrKwSvi9:-NFFJBaB10UhzrKwSviA:b-uvz4ig) ([raw file](https://github.com/cockroachdb/cockroach/blob/0433997b78d3afc029ce01660755192b3ec325c6/pkg/sql/crdb_internal.go#L517)):*
> ```Go
> 		},
> 	},
> 	populate: func(ctx context.Context, p *planner, db catalog.DatabaseDescriptor, addRow func(...tree.Datum) error) error {
> ```

it used to use a generator but now it uses populate. is that on purpose? i thought generators were preferred
___
*[`pkg/sql/crdb_internal.go` line 518 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFIfGj1mLEn4bo0J_g:-NFFIfGj1mLEn4bo0J_h:bzchws0) ([raw file](https://github.com/cockroachdb/cockroach/blob/0433997b78d3afc029ce01660755192b3ec325c6/pkg/sql/crdb_internal.go#L518)):*
> ```Go
> 	},
> 	populate: func(ctx context.Context, p *planner, db catalog.DatabaseDescriptor, addRow func(...tree.Datum) error) error {
> 		all, err := p.Descriptors().GetAllDescriptors(ctx, p.txn)
> ```

should this be using the new cluster setting to only lookup descriptors from the current DB if possible?
___
*[`pkg/sql/crdb_internal.go` line 582 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFJ2V_92dHJYbNjMcl:-NFFJ2V_92dHJYbNjMcm:bisnwkv) ([raw file](https://github.com/cockroachdb/cockroach/blob/0433997b78d3afc029ce01660755192b3ec325c6/pkg/sql/crdb_internal.go#L582)):*
> ```Go
> 	ctx context.Context, p *planner, db catalog.DatabaseDescriptor, addRow func(...tree.Datum) error,
> ) (bool, error) {
> 	dbDescs, err := p.Descriptors().GetAllDescriptorsForDatabase(ctx, p.Txn(), db)
> ```

does this need to be guarded by the cluster setting?
___
*[`pkg/sql/resolver.go` line 152 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFI-xQDiZ9A7VlRY28:-NFFI-xQDiZ9A7VlRY29:bu9bn4k) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ba39c4a53c20ba112715166de8f01cf0b98244/pkg/sql/resolver.go#L152)):*
> ```Go
> 	}
> 	flags := p.ObjectLookupFlags(true /* required */, false /* requireMutable */)
> 	flags.ParentID = dbDesc.GetID()
> ```

i didn't realize that we had to do this. why do we?
___
*[`pkg/sql/resolver.go` line 174 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFIGLa6seJOoizaGsv:-NFFIGLa6seJOoizaGsw:b-x4rgqw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ba39c4a53c20ba112715166de8f01cf0b98244/pkg/sql/resolver.go#L174)):*
> ```Go
> 	}
> 	if schemaDesc.SchemaKind() != catalog.SchemaVirtual {
> 		if dbDesc.GetID() != tableDesc.GetParentID() {
> ```

is this even possible, based on the lookup flags above?
___
*[`pkg/sql/scheduledlogging/captured_index_usage_stats.go` line 219 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFK1HoD-fDYklWt4tg:-NFFK1HoD-fDYklWt4th:b-wzz4ol) ([raw file](https://github.com/cockroachdb/cockroach/blob/77a99e817e0c1f1a4e0a224d0a288ca5d135ef7f/pkg/sql/scheduledlogging/captured_index_usage_stats.go#L219)):*
> ```Go
> 			sessiondata.InternalExecutorOverride{
> 				User:     username.RootUserName(),
> 				Database: string(databaseName),
> ```

FWIW, i think making the query use `databaseName.pg_catalog.pg_class` (or `databaseName.crdb_internal.tables` as it did before) would achieve the same result, so i'm not convinced we need this last commit. did you observe differently?

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90116)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @chengxiong-ruan)

___
*[`pkg/sql/crdb_internal.go` line 518 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFIfGj1mLEn4bo0J_g:-NFFKjBw8nn4s2BXENI1:b-7idcy7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0433997b78d3afc029ce01660755192b3ec325c6/pkg/sql/crdb_internal.go#L518)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should this be using the new cluster setting to only lookup descriptors from the current DB if possible?
</blockquote></details>

specifically, if `db != nil`

**Review by ajwerner — COMMENTED** [review]
TFTR

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-:-NFFPAoL9oTjkmXw0ue2:bb73zsl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan and @rafiss)

___
*[`-- commits` line 4 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFJREG-rwTRfqrYmsS):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

not entirely accurate: we support queries like `select * from "".information_schema.tables`
</blockquote></details>

Good point. We only support it for `information_schema`, not `pg_catalog`.

```
demo@127.0.0.1:26257/movr> select * from "".pg_catalog.pg_class;
ERROR: cannot access virtual schema in anonymous database
SQLSTATE: 42704
HINT: verify that the current database is set
```
___
*[`.ijwb/.idea/modules.xml` line 1 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFF9lqAExNcG8_puDkP:-NFFNJE5C2i5FECie9O5:b32wsm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ba39c4a53c20ba112715166de8f01cf0b98244/.ijwb/.idea/modules.xml#L1)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

was this committed on purpose?
</blockquote></details>

no
___
*[`pkg/sql/crdb_internal.go` line 517 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFJBaB10UhzrKwSvi9:-NFFNKCt3oXFI3IyLTlT:b-c1128z) ([raw file](https://github.com/cockroachdb/cockroach/blob/0433997b78d3afc029ce01660755192b3ec325c6/pkg/sql/crdb_internal.go#L517)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

it used to use a generator but now it uses populate. is that on purpose? i thought generators were preferred
</blockquote></details>

I thought populate was preferred. There's some logic which prevents use of the index if you use `generator` here: https://github.com/cockroachdb/cockroach/blob/72735b8229904008f87b294b98edfc1bc38ee2eb/pkg/sql/virtual_schema.go#L561

I don't really get it, but an index join will happen if you use populate and not generator. Neither are commented as being preferred over the other. Who knows about this stuff?
___
*[`pkg/sql/crdb_internal.go` line 518 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFIfGj1mLEn4bo0J_g:-NFFNf-M7Rwc7eT9lnJD:bfr84m8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0433997b78d3afc029ce01660755192b3ec325c6/pkg/sql/crdb_internal.go#L518)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

specifically, if `db != nil`
</blockquote></details>

It can't, sadly. This table includes dropped tables which cannot be found by scanning namespace.
___
*[`pkg/sql/crdb_internal.go` line 582 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFJ2V_92dHJYbNjMcl:-NFFNsPZ5-kvBTJE41_R:b-klnse0) ([raw file](https://github.com/cockroachdb/cockroach/blob/0433997b78d3afc029ce01660755192b3ec325c6/pkg/sql/crdb_internal.go#L582)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

does this need to be guarded by the cluster setting?
</blockquote></details>

Yeah, I suppose so. Another option I can do is disable the index or something. 
___
*[`pkg/sql/resolver.go` line 152 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFI-xQDiZ9A7VlRY28:-NFFO53HE50IUyZZaidT:b206ny4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ba39c4a53c20ba112715166de8f01cf0b98244/pkg/sql/resolver.go#L152)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i didn't realize that we had to do this. why do we?
</blockquote></details>

I'll add commentary to the code. This flag is new. The essence of it is when we had all descriptors, there's logic in the descs.Collection which says that if you ask for an ID and it is not known, then we know it doesn't exist. When I first wrote this change, it made the ORM queries in rttanalysis actually much slower because they resolve a bunch of hashed OIDs which don't exist. We now had to do a lookup for each one. This now allows us to avoid that. 
___
*[`pkg/sql/resolver.go` line 174 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFIGLa6seJOoizaGsv:-NFFOgHJ07Wgyf188pn7:b-ksq637) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ba39c4a53c20ba112715166de8f01cf0b98244/pkg/sql/resolver.go#L174)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is this even possible, based on the lookup flags above?
</blockquote></details>

It ought not be. I worry about cases where we're mixing cached and uncached descriptors. I'll remove it. 
___
*[`pkg/sql/scheduledlogging/captured_index_usage_stats.go` line 219 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFK1HoD-fDYklWt4tg:-NFFP1nW1RVMv-Rz3x5r:b-x22r09) ([raw file](https://github.com/cockroachdb/cockroach/blob/77a99e817e0c1f1a4e0a224d0a288ca5d135ef7f/pkg/sql/scheduledlogging/captured_index_usage_stats.go#L219)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

FWIW, i think making the query use `databaseName.pg_catalog.pg_class` (or `databaseName.crdb_internal.tables` as it did before) would achieve the same result, so i'm not convinced we need this last commit. did you observe differently?
</blockquote></details>

`databaseName.crdb_internal.tables` ignores `databaseName` and shows all tables in all databases. I assumed that that was well-known. It's totally horrific and something I don't think we can change.

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90116)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @chengxiong-ruan)

___
*[`pkg/sql/logictest/logictestbase/logictestbase.go` line 454 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFQeeWAM-MLoDcl8-r:-NFFQeeWAM-MLoDcl8-s:bzargky) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ba39c4a53c20ba112715166de8f01cf0b98244/pkg/sql/logictest/logictestbase/logictestbase.go#L454)):*
> ```Go
> 	},
> 	{
> 		Name:                "local-use-db-lookup",
> ```

oh missed this earlier. instead of this, we could make the cluster setting default be a metamorphic constant. we do that for a few others. (idea would be to save a bit of test overhead)
___
*[`pkg/sql/scheduledlogging/captured_index_usage_stats.go` line 219 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFK1HoD-fDYklWt4tg:-NFFQR4bAJZj4UYFaLrO:breamn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/77a99e817e0c1f1a4e0a224d0a288ca5d135ef7f/pkg/sql/scheduledlogging/captured_index_usage_stats.go#L219)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`databaseName.crdb_internal.tables` ignores `databaseName` and shows all tables in all databases. I assumed that that was well-known. It's totally horrific and something I don't think we can change. 
</blockquote></details>

i didn't know or i forgot. so sad.
___
*[`pkg/bench/rttanalysis/testdata/benchmark_expectations` line 72 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFRAp09d3F-Zk_5lBH:-NFFRAp09d3F-Zk_5lBI:b-b0hfpl) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ba39c4a53c20ba112715166de8f01cf0b98244/pkg/bench/rttanalysis/testdata/benchmark_expectations#L72)):*
> ```
> 4,ORMQueries/information_schema._pg_index_position
> 3,ORMQueries/pg_attribute
> 3,ORMQueries/pg_class
> ```

is it surprising that some of these increased? is it worth knowing why?

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90116)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan and @rafiss)

___
*[`pkg/bench/rttanalysis/testdata/benchmark_expectations` line 72 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFRAp09d3F-Zk_5lBH:-NFFSG209zAo2qAwn-cM:bmguz8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ba39c4a53c20ba112715166de8f01cf0b98244/pkg/bench/rttanalysis/testdata/benchmark_expectations#L72)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is it surprising that some of these increased? is it worth knowing why?
</blockquote></details>

It's completely expected. We're now doing an index-join -- a scan on the namespace table followed by a batch of point lookups against descriptor. Before we were just scanning all of descriptor.

**Review by ajwerner — COMMENTED** [review]
I think I've addressed the feedback. PTAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-:-NFFycY77raPkUaWLQyw:b-rqnm5x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan and @rafiss)

___
*[`pkg/sql/crdb_internal.go` line 582 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFJ2V_92dHJYbNjMcl:-NFFyMKq5FA8SdUXXP66:b20fbai) ([raw file](https://github.com/cockroachdb/cockroach/blob/0433997b78d3afc029ce01660755192b3ec325c6/pkg/sql/crdb_internal.go#L582)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Yeah, I suppose so. Another option I can do is disable the index or something. 
</blockquote></details>

Doing something fancy felt not worth it. 
___
*[`pkg/sql/resolver.go` line 185 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFoEzRBCfpGKiF_WHN:-NFFoEzRBCfpGKiF_WHO:b-h0tyg4) ([raw file](https://github.com/cockroachdb/cockroach/blob/77a99e817e0c1f1a4e0a224d0a288ca5d135ef7f/pkg/sql/resolver.go#L185)):*
> ```Go
> 			return true, true, nil
> 		}
> 	}
> ```

@rafiss, while we're here, how come we don't need to check whether the user has privileges on the schema?



_Code quote:_
```Go
	for scName, ok := iter.Next(); ok; scName, ok = iter.Next() {
		if schemaDesc.GetName() == scName {
			return true, true, nil
		}
	}
```
___
*[`pkg/sql/catalog/descs/collection.go` line 538 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFu6fc2ka6wJdeZ5wi:-NFFu6fc2ka6wJdeZ5wj:b-9ov4g2) ([raw file](https://github.com/cockroachdb/cockroach/blob/77a99e817e0c1f1a4e0a224d0a288ca5d135ef7f/pkg/sql/catalog/descs/collection.go#L538)):*
> ```Go
> 	ctx context.Context, txn *kv.Txn, db catalog.DatabaseDescriptor,
> ) ([]catalog.TableDescriptor, error) {
> 	all, err := tc.GetAllDescriptorsForDatabase(ctx, txn, db)
> ```

I had to revert this: https://github.com/cockroachdb/cockroach/issues/90655
___
*[`pkg/sql/scheduledlogging/captured_index_usage_stats.go` line 219 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFK1HoD-fDYklWt4tg:-NFFnMZS495lGXQrzHFJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/77a99e817e0c1f1a4e0a224d0a288ca5d135ef7f/pkg/sql/scheduledlogging/captured_index_usage_stats.go#L219)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i didn't know or i forgot. so sad.
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/logictestbase/logictestbase.go` line 454 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFQeeWAM-MLoDcl8-r:-NFFnLdPCRbUs9_T_i2s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ba39c4a53c20ba112715166de8f01cf0b98244/pkg/sql/logictest/logictestbase/logictestbase.go#L454)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

oh missed this earlier. instead of this, we could make the cluster setting default be a metamorphic constant. we do that for a few others. (idea would be to save a bit of test overhead)
</blockquote></details>

Done.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90116)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @chengxiong-ruan)

___
*[`pkg/sql/crdb_internal.go` line 517 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFJBaB10UhzrKwSvi9:-NFG9hj_BvY6fu1COBQK:bswvo0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/0433997b78d3afc029ce01660755192b3ec325c6/pkg/sql/crdb_internal.go#L517)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I thought populate was preferred. There's some logic which prevents use of the index if you use `generator` here: https://github.com/cockroachdb/cockroach/blob/72735b8229904008f87b294b98edfc1bc38ee2eb/pkg/sql/virtual_schema.go#L561

I don't really get it, but an index join will happen if you use populate and not generator. Neither are commented as being preferred over the other. Who knows about this stuff?
</blockquote></details>

i think jordan knows. perhaps people on thee queries team do too. anyway, populate sounds fine for this

**Review by rafiss — APPROVED** [review]
lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-:-NFGB3a4AJLuw9v4uV3q:b-7nbn20)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @chengxiong-ruan)

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90116)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @jordanlewis, and @rafiss)

___
*[`pkg/sql/crdb_internal.go` line 517 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/90116#-NFFJBaB10UhzrKwSvi9:-NFKV5iT0c2UlJIVZbWS:bw7o556) ([raw file](https://github.com/cockroachdb/cockroach/blob/0433997b78d3afc029ce01660755192b3ec325c6/pkg/sql/crdb_internal.go#L517)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think jordan knows. perhaps people on thee queries team do too. anyway, populate sounds fine for this
</blockquote></details>

Maybe it's a problem because `populate` gets called later and runs concurrently and so it ought not access descriptors which have not already been accessed? But wouldn't that be a problem in general with these virtual indexes and tables? @jordanlewis can chime in or redirect on this one? The thing is, my change is definitely not making anything worse, I'm just not clear on the contract regarding concurrency? Do we potentially have some races here where `SELECT * FROM crdb_internal.tables UNION ALL SELECT * FROM crdb_internal.tables` will potentially result in a race condition?

**Comment by jordanlewis** [comment]
Generator is preferred. Populate has the problem that it materializes everything in memory first. I didn't read the full PR here so I could be missing some of the context for this question.

**Comment by ajwerner** [comment]
> Generator is preferred. Populate has the problem that it materializes everything in memory first. I didn't read the full PR here so I could be missing some of the context for this question.

Interesting. I think this means I'm fixing a bug due to illegal concurrency. I imagine the returned callback is run asynchronously? @jordanlewis the problem I encountered is that a virtual index was not being chosen if there was a generator but is chosen if the main table uses populate. @mgartner are you the person to bring this up with?

**Comment by ajwerner** [comment]
Okay, so the problematic lines are here:
https://github.com/cockroachdb/cockroach/blob/accea9d27aeb5afe4d05b23323b8d4b1fac04c77/pkg/sql/virtual_schema.go#L253-L263

These lines are from https://github.com/cockroachdb/cockroach/pull/80422

Why would we want to use a `generator` over a virtual index in either of these cases? Like if it's partial, who cares? Similarly, if there's going to be more than one row, who cares?

**Comment by ajwerner** [comment]
@AlexTalks maybe you recall the motivation for these limitations?

**Comment by jordanlewis** [comment]
I think the generator vs populate and virtual index vs not questions are orthogonal, right? I think if you have a 1-row returning virtual index I agree you can just use populate as it's easier to implement. But if a virtual index is going to return many rows, you may want to use generate to avoid wasting memory.

The partial thing is important for this reason:

```
	// partial is true if the virtual index isn't able to satisfy all constraints.
	// For example, the pg_class table contains both indexes and tables. Tables
	// can be looked up via a virtual index, since we can look up their descriptor
	// by their ID directly. But indexes can't - they're hashed identifiers with
	// no actual index. So we mark this index as partial, and if we get no match
	// during populate, we'll fall back on populating the entire table.
```

**Comment by ajwerner** [comment]
Alright I added a pair of commits to rename `partial` to `incomplete` and to put in an assertion that `incomplete` indexes not be combined with `generator`. Do you see a problem with the use of `populate` here?

**Comment by jordanlewis** [comment]
> Is it safe to call GetAllDescriptors inside a generator callback?

In theory, IIRC, the generator callback has been carefully designed to have 0 true concurrency. If this doesn't actually work, I agree we would have issues. The goal of the whole generator business is to make it easier to write these really stateful iterators using the Go stack rather than having to hand-assemble a giant confusing stateful mess for every table. But, it's key to your point that they don't actually have concurrency or else lots of things would break.

So if we investigate something it would be making sure that the generator callback isn't actually executing concurrently with anything.

> For the returning of many rows, maybe yes, maybe no.

I think in this particular case using populate sounds fine. The flexibility is there for cases that might need to return many more rows without memory monitoring. I guess we should add some more clarity to the doc strings about when to use which.

> rename partial to incomplete 

Sounds good.


>  Do you see a problem with the use of populate here?
No

**Comment by ajwerner** [comment]
> In theory, IIRC, the generator callback has been carefully designed to have 0 true concurrency. 

This is great to hear!

Flakes were https://github.com/cockroachdb/cockroach/pull/90731 and https://github.com/cockroachdb/cockroach/issues/90730

bors r+

**Comment by rafiss** [comment]
for this and for https://github.com/cockroachdb/cockroach/pull/90597, could we add some rttanalysis benchmark that shows fewer lookups happening?

**Comment by ajwerner** [comment]
> for this and for https://github.com/cockroachdb/cockroach/pull/90597, could we add some rttanalysis benchmark that shows fewer lookups happening?

We cannot show that fewer lookups happen because more lookups happen in terms of round-trips :). The problem is that before we were doing a scan of all descriptors. I think one possibility is to extend the rttanalysis framework to include some sort of counter for number descriptors added to the collection or something. I filed https://github.com/cockroachdb/cockroach/issues/90960

**Comment by ajwerner** [comment]
The first commit is https://github.com/cockroachdb/cockroach/pull/90962. I added a new, important second commit. before the second commit was added, partial predicates on virtual indexes were totally ignored. This is now fixed. The testing for it really comes in the 4th commit. Somebody from @cockroachdb/sql-queries ought to sign off on that small change before I merge this.

**Comment by ajwerner** [comment]
Sadly my simplified attempt to add the partial index predicates for virtual indexes does not quite work because we hide the fact that we might fall back to the primary index underneath the `execbuilder`. We need a way to detect this fallback and then further filter the rows based on the predicate value. In theory, this shouldn't be too hard. We'll want to do it around here:
https://github.com/cockroachdb/cockroach/blob/27a65cca62f48ec35ff0e6877a216dae6cfa2df2/pkg/sql/virtual_schema.go#L713-L720

I don't quite know how to properly plumb the row values in so that we can evaluate the constraint reasonably.

**Comment by ajwerner** [comment]
Alright, I have a fixup commit which does what I suggested and it seems to work, but might be ugly and under-tested.

File: pkg/sql/opt_catalog.go [hunk 3]
```diff
@@ -2429,6 +2429,10 @@ func (oi *optVirtualIndex) InvertedColumn() cat.IndexColumn {
 
 // Predicate is part of the cat.Index interface.
 func (oi *optVirtualIndex) Predicate() (string, bool) {
+	if oi.idx != nil {
```
Comment (mgartner) [inline pkg/sql/opt_catalog.go:2432]: nit: putting the `if oid.idx == nil { ... }` case first would match the other methods and be easier to read IMO

File: pkg/sql/virtual_schema.go [hunk 4]
```diff
@@ -698,16 +743,72 @@ func (e *virtualDefEntry) makeConstrainedRowsGenerator(
 			newConstraint.Spans.Append(idxConstraint.Spans.Get(currentSpan))
 		}
 
+		// We'd want to walk the expression, figure out the columns we need, then
```
Comment (mgartner) [inline pkg/sql/virtual_schema.go:759]: nit: end of the sentence is missing.

File: pkg/sql/virtual_schema.go [hunk 5]
```diff
@@ -698,16 +743,72 @@ func (e *virtualDefEntry) makeConstrainedRowsGenerator(
...
+) (func(datums tree.Datums) (matched bool, _ error), error) {
+	if index == nil {
+		return nil, nil
+	}
+	pred := index.GetPredicate()
+	if pred == "" {
+		return nil, nil
+	}
+	publicColumns := e.desc.PublicColumns()
+	exprs, _, err := schemaexpr.MakePartialIndexExprs(
```
Comment (mgartner) [inline pkg/sql/virtual_schema.go:797]: nit: make a function that can make a single partial index expression, or a leave a TODO for me here

**Review by mgartner — APPROVED** [review]
The **sql,opt: add support for partial indexes on virtual tables** commit LGTM.

File: pkg/sql/virtual_schema.go (hunk 5 above)
Comment (ajwerner) [reply pkg/sql/virtual_schema.go]: Done.

File: pkg/sql/virtual_schema.go (hunk 4 above)
Comment (ajwerner) [reply pkg/sql/virtual_schema.go]: Cleaned this commentary up.

File: pkg/sql/opt_catalog.go (hunk 3 above)
Comment (ajwerner) [reply pkg/sql/opt_catalog.go]: Done.

**Review by ajwerner — COMMENTED** [review]
I reworked the `schemaexpr` stuff to have a cleaner single-expression function and to share logic. @mgartner can you give the last revision a quick peek before I bors it?

**Pull request merged**